otp-authentication/otp-authentication.js

```diff
@@ -0,0 +1,188 @@
+'use latest'
+const express       = require('express');
+const Webtask       = require('webtask-tools');
+const { fromEvent } = require('graphcool-lib');
+const bodyParser    = require('body-parser');
+const stream        = require('stream');
+const base32        = require('thirty-two');
+const crypto        = require('crypto');
+const request       = require('request');
+const speakeasy     = require('speakeasy');
+const qr            = require('node-qr-image');
+const EasyPbkdf2    = require('easy-pbkdf2');
+
+const app = express();
+
+app.use(bodyParser.json());
+
+// Add graphcool api client object to req
+app.use((req, res, next) => {
+  if (req.body.context.graphcool) {
+    req.gc = fromEvent(req.body)
+  };
+  next()
+});
+
+app.post('/registerUser', (req,res) => {
+  // { data: { email: "...", password: "..." } }
+  var easyPbkdf2 = new EasyPbkdf2({"DEFAULT_HASH_ITERATIONS": 10000});
+  var salt = easyPbkdf2.generateSalt(); // `salt` should be treated as opaque, as it captures iterations
+  var password = req.body.data.password;
+  easyPbkdf2.secureHash(password, salt, (err, passwordHash, originalSalt) => {
+    const concatPassword = `${originalSalt}.${passwordHash}`
+    req.gc.api('simple/v1')
+      .request(
+        `mutation ($email: String!, $password: String!) { createUser(email: $email, password: $password) { id } }`,
+        { email: req.body.data.email, password: concatPassword })
+      .then(data => {
+        res.json({data: { id: data.createUser.id }})
+      })
+      .catch(err => res.json({error: err}))
+    });
+})
+
+app.post('/authenticateUser', (req,res) => {
+  // { data: { email: "...", password: "..." } }
+  // { data: { email: "...", password: "...", code: "..." } }
+  req.gc.api('simple/v1')
+    .request(
+      `query ($email: String!) { User(email: $email) { id password isOtpEnabled otpSecret } }`,
+      { email: req.body.data.email })
+    .then(data => {
+      const passwordComponents = data.User.password.split('.')
+      const salt = `${passwordComponents[0]}.${passwordComponents[1]}`
+      const passwordHash = passwordComponents[2]
+
+      var easyPbkdf2 = new EasyPbkdf2({"DEFAULT_HASH_ITERATIONS": 10000});
+      easyPbkdf2.verify(salt, passwordHash, req.body.data.password, (err, valid) => {
+        if (err) res.json({error: err})
+        if (valid) {
+          if (data.User.isOtpEnabled) {
+            if (!req.body.data.code) {
+                res.json({data: { preAuthenticated: true }})
+            }
+            else {
+              const verified = verifyToken(data.User.otpSecret, req.body.data.code.split(/[\\(\\)]/g)[1])
```

Comment:
This can use `req.body.data.code` directly when https://github.com/graphcool/api-bugs/issues/202 is fixed.